litestar/contrib/pydantic/pydantic_dto_factory.py

```diff
@@ -72,8 +74,18 @@ def generate_field_definitions(
             field_definition = model_field_definitions[field_name]
 
             # downtype pydantic types for validation in msgspec
-            if field_definition.annotation in PYDANTIC_TYPE_MAP:
-                open_api_type = PYDANTIC_TYPE_MAP[field_definition.annotation].type
+            #
+            # If an annotation is a specialized generic, e.g., `Json[dict[str, Any]]`, then
+            # field_definition.annotation will be `dict[str, Any]` and the metadata will be
+            # `(Json(),)` - that is, the metadata will be an _instance_ of `Json`, not the
+            # type itself. So we check for that case and use the type of the metadata instead
+            # of the annotation, if it exists in the `PYDANTIC_TYPE_MAP`.
+            annotation = next(
+                (type(t) for t in field_definition.metadata if type(t) in PYDANTIC_TYPE_MAP),
+                field_definition.annotation,
+            )
+            if annotation in PYDANTIC_TYPE_MAP:
+                open_api_type = PYDANTIC_TYPE_MAP[annotation].type
```

Comment:
In pydantic v2, if you have a model like

```py
class MyModel(BaseModel):
    data: Json[dict[str, Any]]
```
The _actual_ annotation that gets added to the model is
```py
class MyModel(BaseModel):
    data: Annotated[dict[str, Any], Json()]
```
You can see these shenanigans here: https://github.com/pydantic/pydantic/blob/4dd40aec7397418e37b1029a0496b486adecbb7c/pydantic/types.py#L1375-L1377*